### PR TITLE
Fix: Prevent buildObject to create recursive call without base condition

### LINKED ISSUE #370: Define nested `Type.Object` in `Type.Record` with typebox result in maximum call stack size exceeded error.
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.24.0

### Node.js version

17.0.1

### Operating system

macOS

### Operating system version

macOS Monterey 12.0.1

### Description

Define nested `Type.Object` in `Type.Record` with [typebox](https://github.com/sinclairzx81/typebox) result in maximum call stack size exceeded error.

### Steps to Reproduce

Below is my dependencies.

```json
{
  "dependencies": {
    "@sinclair/typebox": "^0.21.2",
    "ajv": "^8.8.2",
    "ajv-i18n": "^4.2.0",
    "fast-json-stringify": "^2.7.12",
    "fastify": "^3.24.0"
  }
}
```

 I defined the schema with [typebox](https://github.com/sinclairzx81/typebox).

```javascript
const dictValueSchema = Type.Object({
  text: Type.String(),
  color: Type.String(),
});

const publicDictsResponseSchema = Type.Object({
  genders: Type.Record(Type.String(), dictValueSchema),
  roles: Type.Record(Type.String(), dictValueSchema),
});
```

complied to below json schema:

```json
{
  "type": "object",
  "properties": {
    "genders": {
      "type": "object",
      "patternProperties": {
        "^.*$": {
          "type": "object",
          "properties": {
            "text": {
              "type": "string"
            },
            "color": {
              "type": "string"
            }
          },
          "required": [
            "text",
            "color"
          ]
        }
      }
    },
    "roles": {
      "type": "object",
      "patternProperties": {
        "^.*$": {
          "type": "object",
          "properties": {
            "text": {
              "type": "string"
            },
            "color": {
              "type": "string"
            }
          },
          "required": [
            "text",
            "color"
          ]
        }
      }
    }
  },
  "required": [
    "genders",
    "roles"
  ]
}
```

❌ then I got the error response.

```json
{
  "statusCode": 500,
  "error": "Internal Server Error",
  "message": "Maximum call stack size exceeded"
}
```

But if I changed the `dictValueSchema` to `Type.Any()`, everything is ok.

```javascript
const publicDictsResponseSchema = Type.Object({
  genders: Type.Record(Type.String(), Type.Any()),
  roles: Type.Record(Type.String(), Type.Any()),
});
```

✅ I got the correct response.

```json
{
  "genders": {
    "M": {
      "text": "Male",
      "color": "blue"
    },
    "F": {
      "text": "Female",
      "color": "pink"
    },
    "O": {
      "text": "Other",
      "color": "gray"
    }
  },
  "roles": {
    "USER": {
      "text": "User",
      "color": "blue"
    },
    "ADMIN": {
      "text": "Admin",
      "color": "red"
    }
  }
}
```

Hoping for ur help @mcollina , Thank you very much.

### Expected Behavior

No error, get correct response.

### PR DESCRIPTION
**Pull request by suyash-thakur**
While creating nested JSON schema with typebox  object having patternProperties make buildObject function to create recursive call without any base condition. A simple check that function name is not equal to value retrieved from objectReferenceSerializersMap can solve this issue.

Closes #370 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Eomm** [comment]
Would you mind to add a test?

**Comment by suyash-thakur** [comment]
Would it be ok to import typebox library in test file ? I am unable to reproduce the error without using the library,  Following is the schema generate by typebox -:
```javascript
{
  kind: Symbol(ObjectKind),
  type: 'object',
  properties: {
    genders: {
      kind: Symbol(RecordKind),
      type: 'object',
      patternProperties: [Object]
    },
    roles: {
      kind: Symbol(RecordKind),
      type: 'object',
      patternProperties: [Object]
    }
  },
  required: [ 'genders', 'roles' ]
}
```
The parameters passing to the Symbol function is not string. Copying this schema would throw an error because ObjectKind and RecordKind is not defined. Passing it as string doesn't reproduce the error.

**Comment by Eomm** [comment]
> Would it be ok to import typebox library in test file ?

Yes, if it is a dev-dep it is fine.

> Passing it as string doesn't reproduce the error.

Did you try to use something like: `kind: Symbol.for('foo bar')`?

**Comment by suyash-thakur** [comment]
I have amended my previous commit by adding the test file.

**Review by mcollina — APPROVED** [review]
lgtm

File: test/typebox-nested-schema.test.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+  const nestedSchema = Type.Object({
+    text: Type.String()
+  })
+
+  const RootSchema = Type.Object({
+    object1: Type.Record(Type.String(), nestedSchema),
+    object2: Type.Record(Type.String(), nestedSchema)
+  })
+  console.dir(RootSchema, null)
```
Comment (Eomm) [inline test/typebox-nested-schema.test.js:19]:
```suggestion
```

File: index.js [hunk 2]
```diff
@@ -921,7 +921,7 @@ function toJSON (variableName) {
   `
 }
 
-function buildObject (location, code, name) {
+function buildObject(location, code, name) {
```
Comment (Eomm) [inline index.js:924]: linting

```suggestion
function buildObject (location, code, name) {
```

**Review by Eomm — COMMENTED** [review]
two minor suggestion

**Comment by suyash-thakur** [comment]
Updated the PR. 

- Fixed linting
- Removed Console

**Comment by Eomm** [comment]
CI fails when node 10.

Does typebox support node 10?

**Comment by suyash-thakur** [comment]
I couldn't find any official document which specifies the supported node version. I will check it manually and let you know.

**Comment by suyash-thakur** [comment]
I just checked it manually. Seems like typebox  doesnot support node 10. Map constructor is throwing error in node version 10.
Error -:
```javascript
 schemas = new Map();
            ^
SyntaxError: Unexpected token =
```

**Comment by climba03003** [comment]
> I just checked it manually. Seems like typebox doesnot support node 10. Map constructor is throwing error in node version 10. Error -:
> 
> ```js
>  schemas = new Map();
>             ^
> SyntaxError: Unexpected token =
> ```

The problem is not `Map` constructor, but `public class field` which added on `node@12` or after. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields#public_instance_fields

File: test/typebox-nested-schema.test.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
+'use strict'
+
+const test = require('tap').test
+const build = require('..')
+
+const { Type } = require('@sinclair/typebox')
+
+test('nested object in pattern properties for typebox', (t) => {
```
Comment (climba03003) [inline test/typebox-nested-schema.test.js:8]: Let skip it on `node@10`
```suggestion
test('nested object in pattern properties for typebox', { skip: semver.lt(process.versions.node, '12.0.0') }, (t) => {
```

File: test/typebox-nested-schema.test.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+'use strict'
+
```
Comment (climba03003) [inline test/typebox-nested-schema.test.js:2]:
```suggestion

const semver = require('semver')
```

**Review by Eomm — APPROVED** [review]
LGTM within suggestion and green CI

**Comment by suyash-thakur** [comment]
@climba03003 I did the suggested changes but it looks like the test  is still running for node version 10. Can you help me out with this ?

**Comment by climba03003** [comment]
> @climba03003 I did the suggested changes but it looks like the test is still running for node version 10. Can you help me out with this ?

I have tested on `node@10`, it needs to move the `require` inside the test to prevent it from running the code of `typebox`.

```diff
diff --git a/test/typebox-nested-schema.test.js b/test/typebox-nested-schema.test.js
index 778fb02..816c602 100644
--- a/test/typebox-nested-schema.test.js
+++ b/test/typebox-nested-schema.test.js
@@ -3,9 +3,10 @@
 const test = require('tap').test
 const build = require('..')
 const semver = require('semver')
-const { Type } = require('@sinclair/typebox')

 test('nested object in pattern properties for typebox', { skip: semver.lt(process.versions.node, '12.0.0') }, (t) => {
+  const { Type } = require('@sinclair/typebox')
+
   t.plan(1)

   const nestedSchema = Type.Object({
```

**Comment by mcollina** [comment]
It’s ok to avoid supporting Node v10 if we are targeting only the current main branch of this module. However that would not backport to Fastify v3.

https://github.com/fastify/fast-json-stringify/issues/380

**Comment by climba03003** [comment]
The `git` diff is not very hard to trace what is changed inside this PR.

**Comment by mcollina** [comment]
This PR has a massive amount of unrelated changes (linting?). I would recommend you to start from scratch.

**Comment by VigneshMurugan** [comment]
@mcollina , @suyash-thakur - https://github.com/fastify/fast-json-stringify/pull/387

**Pull request closed without merging**